### DISCUSSION TITLE
@ogma/nestjs-module: importing it globally.

### DISCUSSION THREAD

**Discussion by AdaptingAFM** [jmcdo29/ogma#1121] [Q&A]
Hey, is it possible to import the module globally so I don't have to import and inject it on every module?

**Comment by jmcdo29** [comment]
`OgmaService` should be injectable anywhere once you call `OgmaModule.forRoot/Async()`, though as it's only a single instance you won't be able to make use of different contexts per class unless you add that in yourself.

**Reply by tukusejssirs** [reply]
@jmcdo29, could you explain how can we use `OgmaService` in a service file directly used by root module please?

I have root module in which I import `OgmaModule` like this:

```ts
OgmaModule.forRoot({
  interceptor: {
    http: FastifyParser
  },
  service: {
    application: 'SomeAppName',
    color: true,
    json: false
  }
})
```

Then in the service file (which is added to the `providers` array of the root module) I have `@OgmaLogger(SomeService.name) private readonly logger OgmaService` in `contructor()`.

However, Nest fails that it can’t resolve the `OgmaService`:

```ts
[Nest] 17  - 08/29/2022, 10:37:26 AM   ERROR [ExceptionHandler] Nest can't resolve dependencies of the SomeService (?). Please make sure that the argument OGMA_SERVICE:SomeService at index [0] is available in the SomeModule context.

Potential solutions:
- If OGMA_SERVICE:SomeService is a provider, is it part of the current SomeModule?
- If OGMA_SERVICE:SomeService is exported from a separate @Module, is that module imported within SomeModule?
  @Module({
    imports: [ /* the Module containing OGMA_SERVICE:SomeService */ ]
  })

Error: Nest can't resolve dependencies of the SomeService (?). Please make sure that the argument OGMA_SERVICE:SomeService at index [0] is available in the SomeModule context.

Potential solutions:
- If OGMA_SERVICE:SomeService is a provider, is it part of the current SomeModule?
- If OGMA_SERVICE:SomeService is exported from a separate @Module, is that module imported within SomeModule?
  @Module({
    imports: [ /* the Module containing OGMA_SERVICE:SomeService */ ]
  })

    at Injector.lookupComponentInParentModules ($repo_root/node_modules/@nestjs/core/injector/injector.js:241:19)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async Injector.resolveComponentInstance ($repo_root/node_modules/@nestjs/core/injector/injector.js:194:33)
    at async resolveParam ($repo_root/node_modules/@nestjs/core/injector/injector.js:116:38)
    at async Promise.all (index 5)
    at async Injector.resolveConstructorParams ($repo_root/node_modules/@nestjs/core/injector/injector.js:131:27)
    at async Injector.loadInstance ($repo_root/node_modules/@nestjs/core/injector/injector.js:57:13)
    at async Injector.loadProvider ($repo_root/node_modules/@nestjs/core/injector/injector.js:84:9)
    at async Promise.all (index 7)
    at async InstanceLoader.createInstancesOfProviders ($repo_root/node_modules/@nestjs/core/injector/instance-loader.js:47:9)
```

**Reply by jmcdo29** [reply]
> @jmcdo29, could you explain how can we use `OgmaService` in a service file directly used by root module please?
> 
> I have root module in which I import `OgmaModule` like this:
> 
> ```ts
> OgmaModule.forRoot({
>   interceptor: {
>     http: FastifyParser
>   },
>   service: {
>     application: 'SomeAppName',
>     color: true,
>     json: false
>   }
> })
> ```
> 
> Then in the service file (which is added to the `providers` array of the root module) I have `@OgmaLogger(SomeService.name) private readonly logger OgmaService` in `contructor()`.
> 
> However, Nest fails that it can’t resolve the `OgmaService`:
> 
> ```ts
> [Nest] 17  - 08/29/2022, 10:37:26 AM   ERROR [ExceptionHandler] Nest can't resolve dependencies of the SomeService (?). Please make sure that the argument OGMA_SERVICE:SomeService at index [0] is available in the SomeModule context.
> 
> Potential solutions:
> - If OGMA_SERVICE:SomeService is a provider, is it part of the current SomeModule?
> - If OGMA_SERVICE:SomeService is exported from a separate @Module, is that module imported within SomeModule?
>   @Module({
>     imports: [ /* the Module containing OGMA_SERVICE:SomeService */ ]
>   })
> 
> Error: Nest can't resolve dependencies of the SomeService (?). Please make sure that the argument OGMA_SERVICE:SomeService at index [0] is available in the SomeModule context.
> 
> Potential solutions:
> - If OGMA_SERVICE:SomeService is a provider, is it part of the current SomeModule?
> - If OGMA_SERVICE:SomeService is exported from a separate @Module, is that module imported within SomeModule?
>   @Module({
>     imports: [ /* the Module containing OGMA_SERVICE:SomeService */ ]
>   })
> 
>     at Injector.lookupComponentInParentModules ($repo_root/node_modules/@nestjs/core/injector/injector.js:241:19)
>     at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
>     at async Injector.resolveComponentInstance ($repo_root/node_modules/@nestjs/core/injector/injector.js:194:33)
>     at async resolveParam ($repo_root/node_modules/@nestjs/core/injector/injector.js:116:38)
>     at async Promise.all (index 5)
>     at async Injector.resolveConstructorParams ($repo_root/node_modules/@nestjs/core/injector/injector.js:131:27)
>     at async Injector.loadInstance ($repo_root/node_modules/@nestjs/core/injector/injector.js:57:13)
>     at async Injector.loadProvider ($repo_root/node_modules/@nestjs/core/injector/injector.js:84:9)
>     at async Promise.all (index 7)
>     at async InstanceLoader.createInstancesOfProviders ($repo_root/node_modules/@nestjs/core/injector/instance-loader.js:47:9)
> ```

Did you use `OgmaModule.forFeature()` to set up the `OgmaService` for `SomeService`? `OgmaModule.forFeature(SomeService)`?

**Reply by tukusejssirs** [reply]
I did try to add `OgmaModule.forFeature(SomeService.name)` right after `OgmaModule.forRoot(...)` (in the same module), however, it didn’t actually worked (same error occurred).

I’d like to emphasise that in my use case, I have a single (root) module in a microservice.

**Reply by jmcdo29** [reply]
Hmm, that error shouldn't occur. Got a reproduction?

**Reply by tukusejssirs** [reply]
I’ll create one.

**Reply by tukusejssirs** [reply]
@jmcdo29, sorry to bothering you with this, but in my MWE all works as expected. Thanks for your time and help! :pray:

**Reply by tukusejssirs** [reply]
Actually, it works when I use `OgmaModule.forFeature(SomeService.name)` in module imports, but fails when I try to use some arbitrary string like `OgmaModule.forFeature('SomeContext')`.

Is that intentional? From `OgmaModule.forFeature` type definition I read that it accepts a `string`, however, it fails unless the string is set to `name` of a service.

Same error happens when I change the parameter of `@OgmaLogger()` decorator: `@OgmaLogger(AppService.name) private readonly logger: OgmaService` works, but `@OgmaLogger('SomeContext') private readonly logger: OgmaService` fails.

**Reply by tukusejssirs** [reply]
Another issue with the same error: I created a MWE in Nx monorepo with two apps: `test-app` (as gateway) and `microservice` (as TCP microservice).

Everything work as expected (expect the strings being not accepted by Ogma as stated above). However, when I added another service file (there are two services in total: `AppService` and `SomeService`), it behaves strangely; see microservice’s `AppModule` snippets below:

```ts
// WORKS

// Module
@Module({
  imports: [
    OgmaModule.forRoot({
      interceptor: {
        http: FastifyParser
      },
      service: {
        application: 'Microservice',
        color: true,
        json: false
      }
    }),
    OgmaModule.forFeature(AppService.name)
  ],
  controllers: [AppController],
  providers: [AppService],
})
export class AppModule {}

///////////////////////////////////

// FAILS

// Module
@Module({
  imports: [
    OgmaModule.forRoot({
      interceptor: {
        http: FastifyParser
      },
      service: {
        application: 'Microservice',
        color: true,
        json: false
      }
    }),
    OgmaModule.forFeature(AppService.name)
  ],
  controllers: [AppController],
  providers: [AppService, SomeService],
})
export class AppModule {}

///////////////////////////////////

// FAILS

// Module
@Module({
  imports: [
    OgmaModule.forRoot({
      interceptor: {
        http: FastifyParser
      },
      service: {
        application: 'Microservice',
        color: true,
        json: false
      }
    }),
    OgmaModule.forFeature(SomeService.name)
  ],
  controllers: [AppController],
  providers: [AppService, SomeService],
})
export class AppModule {}

///////////////////////////////////


```

All failing example fail with the same error as I stated in [this comment](https://github.com/jmcdo29/ogma/discussions/1121#discussioncomment-3499263) (replace `SomeService` with `AppService` and `SomeModule` with `AppModule`; `SomeService` is never in the error message).

---

It sometimes works when I use a Ogma only in one of the services only, but not always.

---

MWE/reproduction: https://gitlab.com/tukusejssirs/nx-ogma

After `npm ci`, `npx nx serve microservice` fails.

**Reply by jmcdo29** [reply]
Interesting. I'll definitely be taking a look later

**Reply by jmcdo29** [reply]
Where is the `OgmaModule.forFeature(AppService.name)`? You use both `@OgmaLogger(AppService.name)` and `@OgmaLogger(SomeService.name)`, so you need both `forFeature` imports, or to use `OgmaModule.forFeatures([AppService.name, SomeService.name])`

**Reply by tukusejssirs** [reply]
Argh, okay. :pray:

I didn’t know of `forFeatures()` method.

However, could you explain me please why do we need to specify all features in a module? Can’t we simply use `@OgmaLogger(SomeService.name)` (defined in constructor of a service) without the need to specify what features we are able to use in a particular module?

**Reply by jmcdo29** [reply]
This has to do with how I decided to write the `OgmaModule`. So each `@OgmaModule()` is really an alias for `@Inject('OGMA_SERVICE:token')`. I don't do anything like adding the tokens to an array that gets populated at start time, as I've seen described in some articles, I just make the decorator a simple alias to help keep things simple. So, as `@OgmaLogger()` is just setting up an injection token, we also need to tell Nest about that injection token, and we can do that using the `OgmaModule.forFeature()` dynamic module, so that we can set up loggers per feature that we want to use. If you don't want to add the context per logger, you could just inject `OgmaService` into your constructor and be done with it, but I've found the context per class to be helpful. 

I designed it this way to give more clarity into what is being used in the module and less hand-waviness behind the scenes. I generally prefer more explicit config than implicit

**Reply by tukusejssirs** [reply]
Thanks for explanation. :pray: It makes sense, however, it might be a good idea to explain this in the docs and/or `README.md` files. Maybe it’s just be not being able to find what I am looking for. :man_shrugging:

**Reply by jmcdo29** [reply]
First line of [this page](https://ogma.jaymcdoniel.dev/docs/nestjs/service)

> The `@ogma/nestjs-module` comes with its own decorator to be used with `OgmaModule.forFeature()`, very similar to `@nestjs/typeorm` and `@nestjs/mongoose`. To inject an `OgmaService`, simply use the `@OgmaLogger()` decorator and provide the context for the injection

Can I make this clearer somehow?

**Reply by tukusejssirs** [reply]
Well, maybe it’s just me being more of a TS/Nest newbie, but I didn’t get [this](https://github.com/jmcdo29/ogma/discussions/1121#discussioncomment-3510319) out of the quoted text from docs.

Based on what I have learnt from chatting with you in this discussion, I would expect to have the following information to be somewhere in [there](https://ogma.jaymcdoniel.dev/docs/nestjs/service):

- one needs to list all tokens in `OgmaModule.forFeature()` or `OgmaModule.forFeatures()` in module file that can be used in that particular module (in any file within that module);
- `OgmaModule.forFeature()` accepts a single token and `OgmaModule.forFeatures()` accepts an array of tokens;
- [this is already in the docs] a token specified in `@OgmaLogger()` in a constructor of a service can be overrided by specifying a context as the second parameter of the log command (e.g. `this.logger.log('message', {context: 'new-token'}).

**Reply by tukusejssirs** [reply]
@jmcdo29, I am here to ask you if it was possible to add support of `isGlobal` property so that we could make `OgmaModule` a global module and then we would need to define only `OgmaModule.forFeature()`/`OgmaModule.forFeatures()` in other (non-root/sub-) modules.

---

As I see it, we usually want to have a single logger configuration per app. The logger config is usually dependent on some app config variables (e.g. in `.env` we define the wanted log level and format).

Now, when I want to use `@nestjs/config`, I need to add `OgmaModule` asynchronously:

```ts
@Module({
  imports: [
    ConfigModule.forRoot({
      isGlobal: true,
      load: [(): Config => customConfig],
    }),
    OgmaModule.forRootAsync({
      imports: [ConfigModule],
      inject: [ConfigService],
      useFactory: (configService: ConfigService) => ({
        interceptor: {
          http: FastifyParser,
        },
        isGlobal: true,  // Note: This does not work.
        service: {
          application: 'appName',
          color: true,
          each: true,
          json: false,
          logLevel: configService.get<OgmaOptions['logLevel']>('logLevel'),
        },
      }),
    })
  ]
})
```

It works in itself, ~~but `ConfigModule` needs to be imported directly in both root module and any other submodule (otherwise it does not work, even though `ConfigModule` is imported in the root module)~~, which makes it non-DRY. However, I think Ogma users should be able to defined the Ogma logger config once per app, in order to deduplicate the configuration.

In the end, I don’t want to import `ConfigModule` nor `OgmaModule.forRootAsync()` in any of the submodules, as they are already configured in the root module.

**Update:** Actually, it works without `CacheModule` being imported in the submodules, but I need to remove it from the imports of `OgmaModule.forRootAsync()` too, like so:

```ts
@Module({
  imports: [
    OgmaModule.forRootAsync({
      inject: [ConfigService],
      useFactory: (configService: ConfigService) => ({
        interceptor: {
          http: FastifyParser,
        },
        isGlobal: true,  // Note: This does not work.
        service: {
          application: 'appName',
          color: true,
          each: true,
          json: false,
          logLevel: configService.get<OgmaOptions['logLevel']>('logLevel'),
        },
      }),
    })
  ]
})
```

---

**Update 2:** Actually, it seems to work. I missed `ConfigService` provider in the service file test; when I added it, it all works as expected. Sorry for the noise.
